juju/juju
# Remove facade context attributes

Various api related structs stored a context on an attribute. This PR removes said attributes and instead updates the various methods where the context is needed to take it as an argument.

There was a lot of gomock and mechanical test fallout - these are in separate commits.

5007a8745dd1ab600d237add4d1f8e176c23b8c8 is the gomock changes
https://github.com/juju/juju/pull/16756/commits/4c2e81e1f4f86d0833f5642665b67b568e222e15 is the mechanical test changes to add `context.Background()` etc to the api calls being used.

The actual changes are in commit https://github.com/juju/juju/pull/16756/commits/af2138879a6bf161bb49e472d3915bd8e815cb0a.

Note we now start out with a nil tracer on the facade struct - if you want a tracer, you need to use `WithTracer()`. This is the same behaviour as we had since we used to just start with a no op tracer anyway as the now deleted context attribute was just a `context.Background()`.

There's still a lot more to do, but this cleans up a fairly significant wart.

## QA steps

bootstrap
deploy juju-qa-dummy-source and juju-qa-dummy-sink in different models and cross model relate
ensure it all works

## Links

**Jira card:** JUJU-5243
